tools/hsmtool.c

```diff
@@ -368,6 +371,91 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	return 1;
 }
 
+static void read_mnemonic(char *mnemonic) {
+	printf("Introduce the BIP39 24 word list:\n");
+	char word[50];
+	uint n_words;
+	for (n_words = 0; n_words < 24; n_words++) {
+		scanf("%s", word);
+		/* We need to concat the strings using ' ' so libwally can
+		 * validate it */
+		if (n_words == 0) {
+			strcpy(mnemonic, word);
+		} else {
+			strcat(mnemonic, " ");
+			strcat(mnemonic, word);
+		}
+	}
+	/* scanf() leaves the newline character in the input buffer */
+	getchar();
+}
+
+static void read_passphrase(char *passphrase) {
+	struct termios current_term, temp_term;
+	printf("Warning: remember that different passphrase yield different "
+	       "bitcoin wallets.\n");
+	printf("Leave it empty for using the value by default (echo is "
+	       "disabled now).\n");
+	printf("Enter your passphrase: ");
+
+	/* Change terminal options so we do not echo the passphrase */
+	if (tcgetattr(fileno(stdin), &current_term) != 0)
+		errx(ERROR_USAGE, "Could not get current terminal options.");
+	temp_term = current_term;
+	temp_term.c_lflag &= ~ECHO;
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &temp_term) != 0)
+		errx(ERROR_USAGE, "Could not disable passphrase echoing.");
+	/* If we don't flush we might end up being buffered and we might seem
+	 * to hang while we wait for the password. */
+	fflush(stdout);
+
+	size_t passphrase_size;
+	if (getline(&passphrase, &passphrase_size, stdin) < 0)
+		errx(ERROR_USAGE, "Could not read passphrase from stdin.");
+	printf("passphrase size: %ld\n", passphrase_size);
+	/* Replace the last character which will always be '\n' */
+	passphrase[strlen(passphrase) - 1] = '\0';
```

Comment:
Not always:

```C
#include <inttypes.h>
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {
	char *test = NULL;
	size_t len;

	getline(&test, &len, stdin);
	printf("%s\n", test[strlen(test) - 1] == '\n' ? "true" : "false");
}
```

```
darosior@darosior:~$ gcc a.c
darosior@darosior:~$ ./a.out 
test_pass
true
darosior@darosior:~$ printf passphrase_from_elsewhere |./a.out 
false
```

Reply:
This is a really good catch because I did not think about that... Thanks @darosior!

Reply:
Ok, after some time playing with this I think that we have a problem here... I do not see a way to know if the last character is noise or was introduced by the user. The reason is that `getline` returns the number of characters that where read, including the `\n` if present but not putting it in the read bytes. Then, if we know that we readed 2 chars, we do not know if `test[1]` is noise that was there because the user pushed enter `\n` or if the user executed `printf "xc" | ./myprogram` and then test[1] is the "c"

I think that the best is going to be to assume that the user is using this interactively and always replace the last character by `\0` as it was the enter and was not supposed to be part of the passphrase.

If not, we have that


```
#include <inttypes.h>
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {
	char *test = NULL;
	size_t len;

	size_t characters = getline(&test, &len, stdin);
        printf("read chars: %ld\n", characters);
	printf("%s\n", test[strlen(test) - 1] == '\n' ? "true" : "false");
}
```

Then...

```
~/tmp ~$ ./a.out 
a
read chars: 2
true
~/tmp ~$ printf "ab" | ./a.out
read chars: 2
false
```

Reply:
Why not doing it as we do it for `--encrypted-hsm` ?
https://github.com/ElementsProject/lightning/blob/cafaad741b39806a8bd9fff7fe71100488cf4bed/lightningd/options.c#L417-L418